Symlink as workspace root breaks file watching

### Check for existing issues

- [X] Completed

### Describe the feature

Right now, if I make a new file or directory outside of the editor (like by downloading one with `curl`, or if my program makes a config file, etc), it does not appear in the editor's file view until I restart the program. Not sure whether this is a bug or would be a new feature.

If it's a bug, I'm having the issue on Linux, on past versions, and the current one in the Arch repos (0.150.4)

### Environment

Edit: I tried this on two machines, with two different versions of Zed, but I'm now up-to-date on both. The old version (0.149.5) was not the issue, I just hadn't updated on the machine I originally posted this from.

Zed: v0.150.4 (Zed)
OS: Linux Wayland arch 20240721.0.248532
Memory: 31 GiB
Architecture: x86_64
GPU: AMD Radeon RX 6800 (RADV NAVI21) || radv || Mesa 24.2.1-arch1.1

### Zed Log

<details><summary>Zed.log</summary><pre>
with error: failed to connect to the server
2024-09-02T12:32:22.070436438+01:00 [ERROR] ignoring event "/home/user/Downloads/test/test" outside of root path "/home/user/test"
2024-09-02T12:32:25.165355705+01:00 [INFO] Using surface present mode MAILBOX
</pre></details>

I figured out what the problem might be. `fs::watch` indirectly uses `readlink` from libc on Unix. `readlink` only resolves to a value if the path *itself* is a symlink, not one of its parents. `smol::fs::read_link` returns a `Err(smol::io::Error)` if the path is not a symlink but errors get discarded by the `.await.ok()` at `fs.rs:591`.

This means if the workspace root is `~/dir1/dir2/` where `dir1` is a symlink, `target_path` gets set to `None` on line 591. The closure inside `filter_map` on line 608 then returns `None` for every path because `notify::Event`s always contain the canonical path returned from `inotify` so they never start with the non-canonical `root_path`. The `.collect::<Vec<_>>()` produces an empty `Vec`  and `pending_paths` never gets extended.

All notifications from the watcher get rejected because the returned canonical path doesn't match the workspace root inside Zed. This prints out messages like this to the log: `[ERROR] ignoring event "/home/user/dir1/file" outside of root path "/home/user/link1"` where `link1` is a symlink to `dir1`.

This could be fixed by replacing line 591 with:
```rust
let target_path = self.canonicalize(&path).await.log_err();
```
and all uses of `root_path` with the canonical `target_path` .

**Note:** this is all specific to the Linux implementation of `fs::watch`. The macOS and Windows versions don't handle symlinks this way and probably don't have this issue.

--- a/crates/workspace/src/workspace.rs
+++ b/crates/workspace/src/workspace.rs
@@ -1096,10 +1096,17 @@ impl Workspace {
         );
 
         cx.spawn(|mut cx| async move {
-            let serialized_workspace: Option<SerializedWorkspace> =
-                persistence::DB.workspace_for_roots(abs_paths.as_slice());
+            let mut paths_to_open = Vec::with_capacity(abs_paths.len());
+            for path in abs_paths.into_iter() {
+                if let Some(canonical) = app_state.fs.canonicalize(&path).await.ok() {
+                    paths_to_open.push(canonical)
+                } else {
+                    paths_to_open.push(path)
+                }
+            }
 
-            let mut paths_to_open = abs_paths;
+            let serialized_workspace: Option<SerializedWorkspace> =
+                persistence::DB.workspace_for_roots(paths_to_open.as_slice());
 
             let workspace_location = serialized_workspace
                 .as_ref()

